staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/token.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	tokens2 "github.com/gophercloud/gophercloud/openstack/identity/v2/tokens"
+	tokens3 "github.com/gophercloud/gophercloud/openstack/identity/v3/tokens"
+	"github.com/gophercloud/gophercloud/openstack/utils"
+)
+
+const expiresAtDelta = 10 * time.Second
+
+type openstackToken struct {
+	ID        string
+	ExpiresAt time.Time
+}
+
+func (o *openstackToken) Valid() bool {
+	return o != nil && o.ID != "" && !o.expired()
+}
+
+func (o *openstackToken) expired() bool {
+	if o.ExpiresAt.IsZero() {
+		return false
+	}
+	return o.ExpiresAt.Add(-expiresAtDelta).Before(time.Now())
+}
+
+func tokenFromAuth(options gophercloud.AuthOptions) (*openstackToken, error) {
+	client, err := openstack.NewClient(options.IdentityEndpoint)
+	if err != nil {
+		return &openstackToken{}, fmt.Errorf("error creating openstack client: %s", err)
+	}
+
+	versions := []*utils.Version{
+		{ID: "v2.0", Priority: 20, Suffix: "/v2.0/"},
+		{ID: "v3", Priority: 30, Suffix: "/v3/"},
+	}
+
+	chosen, endpoint, err := utils.ChooseVersion(client, versions)
+	if err != nil {
+		return &openstackToken{}, fmt.Errorf("error choosing openstack identity API version: %s", err)
+	}
+
+	var token *openstackToken
+	switch chosen.ID {
+	case "v2.0":
+		token, err = v2auth(client, endpoint, options, gophercloud.EndpointOpts{})
+	case "v3":
+		token, err = v3auth(client, endpoint, &options, gophercloud.EndpointOpts{})
+	default:
+		return &openstackToken{}, fmt.Errorf("unrecognized openstack identity API version: %s", chosen.ID)
+	}
+	if err != nil {
+		return &openstackToken{}, fmt.Errorf("error authenticating with openstack: %s", err)
+	}
+	return &openstackToken{token.ID, token.ExpiresAt}, nil
+}
+
+func v2auth(client *gophercloud.ProviderClient, endpoint string, options gophercloud.AuthOptions, eo gophercloud.EndpointOpts) (*openstackToken, error) {
+	v2Client, err := openstack.NewIdentityV2(client, eo)
+	if err != nil {
+		return nil, err
+	}
+
+	if endpoint != "" {
+		v2Client.Endpoint = endpoint
+	}
+
+	v2Opts := tokens2.AuthOptions{
+		IdentityEndpoint: options.IdentityEndpoint,
+		Username:         options.Username,
+		Password:         options.Password,
+		TenantID:         options.TenantID,
+		TenantName:       options.TenantName,
+		TokenID:          options.TokenID,
+	}
```

Comment:
Thanks, will fix.

I initially thought `v2Opts` is declared because `tokens2.Create` would fail is passed v3-speciifc options like DomainID and DomainName. But now I see that [`ToTokenV2CreateMap` ](https://github.com/gophercloud/gophercloud/blob/d240260ea5048753e7c2f7afffd13ee6979ac179/openstack/identity/v2/tokens/requests.go#L56)takes care of this conversion internally.